### ISSUE TITLE
During cron, use most recently used language for post type archive links

### ISSUE THREAD

**Issue by davelit** [polylang/polylang#828]
## Prerequisites

- [x] This is not a usage question (Those should be directed to the [community supported forum](https://wordpress.org/support/plugin/polylang), unless this is a question about Polylang Pro in which case you should [use the helpdesk](https://polylang.pro/support/)).
- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [x] The issue still exists against the latest `master` branch of Polylang and the latest WordPress version.
- [x] This bug happens with only Polylang plugin active
- [x] This bug happens with a default WordPress theme active
- [x] I can reproduce this bug consistently

## Steps to reproduce the issue

1. Create a custom post type in WordPress
2. In Polylang, set the URL modifications setting to "_The language is set from the directory name in pretty permalinks_" and uncheck the "_Hide URL language information for default language_" option. Then enable translation for the new post type in the "_Custom post types and Taxonomies_" setting.
3. Create a post with the new post type and publish it.
4. Create a WP Cron job that prints the URL of the new post (`get_permalink($post)`) and the URL of the corresponding post type archive (`get_post_type_archive_link($post->post_type)`) to the error_log.

## Expected behavior and actual behavior

When I follow those steps, I see the correct URL for the custom post in the error log, but the language folder is missing from the URL of the corresponding post type archive.

I'd like to see the localized URL of the post type archive that belongs to the custom post.

### Use Case

A lot of sites are using caching plugins like Cache Enabler, WP Rocket, or WP Super Cache to improve their [Core Web Vitals](https://web.dev/vitals/). When a new blog post is published, these plugins offer the option to either clear the whole cache or only associated pages, like previous and next posts as well as term and post type archive pages. The latter option is preferrable on larger sites that are updated several times a day, since it's computationally expensive to regenerate the whole cache and difficult to keep the cache warm if it is wiped every time a new blog post gets pulished.

### Root Problem

Unfortunately, it's difficult for the caching plugins to determine which language's post type archive should actually be cleared if a new post is published. Everything still works if you publish the post right away. In that case, `PLL_Filters_Links->post_type_archive_link()` can determine the language from the current post or the current user's session. But if you schedule the post to be published later, `PLL_Filters_Links->post_type_archive_link()` will be executed in the context of a WP Cron job with no language attached to it (`PLL_Filters_Links->$curlang` will be empty). Consequently, `get_post_type_archive_link()` cannot be localized anymore, and the caching plugins get the wrong URL to purge.

### Suggested Solution

I don't see a clean, straightforward solution to this fundamental problem, but a workaround that would at least serve as a remedy for the abovementioned caching scenario. My suggestion is to keep track of the language that was last used in the context of a post type and to fall back to it if no current language can be determined. 

To give you a better idea, here's what I'm currently using to work around the post type archive link issue on my system (in combination with Cache Enabler):

```php
add_filter( 'pll_post_type_link', function( $link, $lang ) {
    global $polylang_recent_lang;
    if ( ! empty( $lang) && wp_doing_cron() ) {
        $polylang_recent_lang = $lang;
    }
    return $polylang_recent_lang;
}, 10, 2 );


add_filter( 'post_type_archive_link', function( $link ) {
    global $polylang, $polylang_recent_lang;
    if (       ! empty( $polylang ) 
            &&   empty( $polylang->curlang ) 
            && ! empty( $polylang_recent_lang ) 
            &&   wp_doing_cron() ) {
        $polylang->curlang = $polylang_recent_lang;
    }
    return $link;
}, 19);
```
## Thank You

Since this is my first comment on the Polylang project on GitHub, let me use this opportunity to thank you for your work and to commend you on Polylang's exemplary code quality. It's a pleasure to work with! 

I'm responsible for a fairly large site (10k daily visits, 12k posts) and we gladly pay the annual Pro license even though we also have a lifetime WPML license, which we used in the past. We switched because of reliability issues and we wanted the same slugs across all languages (not sure right now, whether WPML has (had) limitations in this regard). We also need per-language user rights and looked into MultilingualPress for this purpose, but didn't switch since that plugin seems to have lower market adoption and we feared that we'd run into more compatibility issues because of this.

**Comment by Chouby** [comment]
It looks strange to me that a cache plugin would use cron to clear the cache when updating a post. Are you sure it's the case?

If I well understood the code of your cache plugin, it looks like the cache is cleared directly when the post is updated: https://github.com/keycdn/cache-enabler/blob/1.7.1/inc/cache_enabler.class.php#L81

I also checked with WP Rocket and could observe the same issue.

**Comment by davelit** [comment]
Thanks for taking the time to look into this. 

My bad, I tried to describe the problem in a plugin agnostic way, but that made the issue look like a weird edge case.

Let me try again by telling you how I built by test case:
1. I created a new blank-slate WordPress install, selected the default theme and installed two plugins: Polylang and Cache Enabler
2. I created a custom plugin that registered a new custom post type called `custom-post-type` (slug)
3. In Polylang, I set the URL modifications setting to "The language is set from the directory name in pretty permalinks" and unchecked the "Hide URL language information for default language" option. I then enabled translations for the new post type in the "Custom post types and Taxonomies" setting.
4. I added an `else` clause to the first `if` statement in `PLL_Links_Directory->add_language_to_link()` with the following contents:
```php
public function add_language_to_link( $url, $lang ) {
    if ( ! empty( $lang ) ) {
        /* ...existing code... */
    } else {
        error_log( "PLL_Links_Directory->add_language_to_link( '$url', null ):\n\t"
                  . str_replace( ', ', ",\n\t", wp_debug_backtrace_summary() ) );
    }
    return $url;
}
```
5. I created a post with the new post type and **scheduled it for publishing in 3 minutes**.
6. I opened my test WordPress installation in an incognito window, waited 3 minutes and then refreshed the page, triggering WordPress' cron job that published the post I had scheduled moments before.
7. I found the following in the error log:
```
PLL_Links_Directory->add_language_to_link( 'https://test.local/custom-post-type/', null ): 
	do_action_ref_array('publish_future_post'),
	WP_Hook->do_action,
	WP_Hook->apply_filters,
	check_and_publish_future_post,
	wp_publish_post,
	do_action('save_post'),
	WP_Hook->do_action,
	Cache_Enabler::on_save_trash_post,
	Cache_Enabler::clear_cache_on_post_save,
	Cache_Enabler::clear_associated_cache,
	Cache_Enabler::clear_post_type_archives_cache,
	get_post_type_archive_link,
	apply_filters('post_type_archive_link'),
	WP_Hook->apply_filters,
	PLL_Filters_Links->post_type_archive_link,
	PLL_Links_Model->switch_language_in_link,
	PLL_Links_Directory->add_language_to_link
```
I hope this makes the issue clearer. Thanks again and let me know if I can help you with the investiagtion in any way.

**Comment by Chouby** [comment]
Coud you tell me if #831 fixes the issue for you?

**Comment by davelit** [comment]
Yes it does! And also fixes an edge case that I didn't see before looking at your new code (language selector in admin interface is set to a different language than the actual post that is published -> PLL()->curlang is admin language, not post language, so post type archive of **admin** language is cleared).

Thank you again for your outstanding work!

**Comment by Chouby** [comment]
Thanks for your feedback.
